java-api/src/main/kotlin/io/spine/validation/api/MessageValidator.kt

```diff
@@ -0,0 +1,129 @@
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.validation.api
+
+import com.google.protobuf.Message
+import io.spine.annotation.SPI
+
+/**
+ * A custom validator for Protobuf messages of type [M] that are defined externally.
+ *
+ * External messages are those that come from dependencies.
+ * Protobuf [well-known](https://protobuf.dev/reference/protobuf/google.protobuf/)
+ * messages should also be considered external. For such messages, it is impossible
+ * to declare validation constraints using the validation options because there's
+ * no access to their proto definitions.
+ *
+ * To be able to validate external messages, one must implement this interface
+ * and annotate the implementing class with the [Validator] annotation, specifying
+ * the type of the message to validate.
+ *
+ * An example of the validator declaration for the `Earphones` message:
+ *
+ * ```kotlin
+ * @Validator(Earphones::class)
+ * public class EarphonesValidator : MessageValidator<Earphones> {
+ *     public override fun validate(message: Earphones): List<DetectedViolation> {
+ *         return emptyList() // Always valid.
+ *     }
+ * }
+ * ```
+ *
+ * ## Applicability
+ *
+ * A validator is applied only to the local messages of the module it is declared in.
+ * For each local message that has a field of type [M], the validation library
+ * will invoke a validator when validating that message.
+ *
+ * The following field types are supported:
+ *
+ * 1. Singular fields of type [M].
+ * 2. Repeated field of type [M].
+ * 3. Map field with values of type [M].
+ *
+ * Please note that standalone instances of [M] and fields of [M] type that occur in
+ * external messages will not be validated using the validator.
+ *
+ * Consider the following example:
+ *
+ * ```proto
+ * import "earphones.proto"; // Brings the `Earphones` message from dependencies.
+ *
+ * // A locally-declared message.
```

Comment:
See, the problem with this "local-external" is that they are not defined properly. Please devote a section in the beginning to the explanations and the definitions.